ynr/apps/sopn_parsing/helpers/extract_pages.py

```diff
@@ -32,3 +40,194 @@ def extract_pages_for_ballot(ballot):
         raise PDFException(
             f"Failed to extract pages for {ballot.sopn.uploaded_file.path} as a PDFSyntaxError was raised"
         )
+
+
+textract_client = boto3.client("textract", region_name="eu-west-2")
+
+
+def start_detection(ballot):
+    """This is Step 1 of the SOPN parsing process using AWS Textract.
+    https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/textract/client/get_document_analysis.html#
+    """
+
+    file_path = ballot.sopn.uploaded_file.path
+    with open(file_path, "rb") as file:
+        file_bytes = bytearray(file.read())
+    region = "eu-west-2"
+    bucket_name = "public-sopns"
+    s3_client = boto3.client("s3", region_name=region)
+    object_key = "test/test_sopn.pdf"
+
+    response = s3_client.put_object(
+        Bucket=bucket_name,
+        Key=object_key,
+        Body=file_bytes,
+    )
+    print(f"Uploaded bytes to s3://{bucket_name}/{object_key}")
+    response = textract_client.start_document_analysis(
+        DocumentLocation={
+            "S3Object": {
+                "Bucket": bucket_name,
+                "Name": object_key,
+            }
+        },
+        FeatureTypes=["TABLES", "FORMS"],
+        OutputConfig={
+            "S3Bucket": "public-sopns",
+            "S3Prefix": "test",
+        },
+    )
+    if response["JobId"]:
+        job_id = response["JobId"]
+        get_job_results(job_id, ballot)
+    else:
+        print("Job failed to start")
+        raise Exception("Job failed to start")
+
+
+def get_job_results(job_id, ballot):
+    """This is Step 2 of the SOPN parsing process using AWS Textract.
+    https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/textract/client/get_document_analysis.html#
+    """
+
+    textract_result = textract_client.get_document_analysis(JobId=job_id)
+    while textract_result["JobStatus"] not in ["SUCCEEDED", "FAILED"]:
+        sleep(5)
+        textract_result = textract_client.get_document_analysis(JobId=job_id)
+
+    if textract_result["JobStatus"] == "SUCCEEDED":
+        TextractResult.objects.update_or_create(
+            official_document=ballot.sopn,
+            defaults={"job_id": job_id, "json_response": textract_result},
+        )
```

Comment:
I've been faffing with textract for another project this week and discovered a bit of a gotcha here. It's possible that the returned result will be truncated. In this case you need to use the next token to get subsequent parts of the response. I didn't really dig into this as I was just trying to hack something together. But the code below did the job. Essentially it copies the initial response, and creates a seperate list of blocks. Goes through all the `NextTokens` and grabs more blocks to shove in the list. Then subs this complete list of blocks into the copy of the original result and returns it. 

```
    if job_status == "SUCCEEDED":
        analysis = copy.deepcopy(result)
        blocks = []
        while True:
            for block in result["Blocks"]:
                blocks.append(block)
            if ("NextToken" not in result.keys()):
                break
            next_token = result["NextToken"]
            result = textract_client.get_document_analysis(
                JobId=start_response["JobId"],
                NextToken=next_token
            )
        analysis.pop("NextToken", None)
        analysis["Blocks"] = blocks
    else:
        analysis = result

    result_json = json.dumps(result, indent=2)
```

Reply:
Nevermind! Figured it out 🙏